File: leetcode/树/100_相同的树.php
<?php

// 给定两个二叉树，编写一个函数来检验它们是否相同。
//
//如果两个树在结构上相同，并且节点具有相同的值，则认为它们是相同的。
//
//示例 1:
//
//输入:       1         1
//          / \       / \
//         2   3     2   3
//
//        [1,2,3],   [1,2,3]
//
//输出: true
//示例 2:
//
//输入:      1          1
//          /           \
//         2             2
//
//        [1,2],     [1,null,2]
//
//输出: false
//示例 3:
//
//输入:       1         1
//          / \       / \
//         2   1     1   2
//
//        [1,2,1],   [1,1,2]
//
//输出: false
//
//来源：力扣（LeetCode）
//链接：https://leetcode-cn.com/problems/same-tree
//著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
//
//


class TreeNode
{
    public $val = null;
    public $left = null;
    public $right = null;

    function __construct($value)
    {
        $this->val = $value;
    }
}

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution
{

    /**
     * 先序遍历双栈判断是否相等
     * @param TreeNode $p
     * @param TreeNode $q
     * @return bool
     */
    function isSameTree($p, $q)
    {
        $pStack = [$p];
        $qStack = [$q];
        while (!empty($pStack) || !empty($qStack)) {
            $pNode = array_pop($pStack);
            $qNode = array_pop($qStack);
            if ($pNode === null && $qNode === null) {
                continue;
            }
            if ($pNode === null || $qNode === null || ($pNode->val !== $qNode->val)) {
                return false;
            }
            // 注意null节点也需要入栈, 否则无法区分 1. 左树 空右树 2. 空左树 右树 是否相等
            $pStack[] = $pNode->right;
            $pStack[] = $pNode->left;
            $qStack[] = $qNode->right;
            $qStack[] = $qNode->left;
        }
        return true;
    }

    /**
     * 最直观的递归算法
     * 截止条件: p === null && q === null
     * 递归公式: (p.left, q.left) (p.right, q.right)
     * @param TreeNode $p
     * @param TreeNode $q
     * @return Boolean
     */
    function isSameTree1($p, $q)
    {
        if ($p === null && $q === null) {
            return true;
        }
        if ($p !== null && $q !== null && $p->val === $q->val) {
            return $this->isSameTree($p->left, $q->left) && $this->isSameTree($p->right, $q->right);
        } else {
            return false;
        }
    }
}

$s = new Solution();

$p = new TreeNode(1);
$p->left = new TreeNode(2);
$p->right = new TreeNode(3);

$q = new TreeNode(1);
$q->left = new TreeNode(2);
$q->right = new TreeNode(3);


var_dump($s->isSameTree($p, $q));  // true

File: leetcode-golang/tree/100_相同的树.go
package tree

func isSameTree(p *TreeNode, q *TreeNode) bool {
	// 结束条件
	if p == nil && q == nil {
		return true
	} else if p == nil || q == nil {
		return false
	} else {
		return p.Val == q.Val && isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
	}
}

// 非递归
func isSameTree2(p *TreeNode, q *TreeNode) bool {
	// 层序遍历
	type s struct {
		p1 *TreeNode
		p2 *TreeNode
	}
	stack := []s{
		{
			p1: p,
			p2: q,
		},
	}
	for len(stack) > 0 {
		newStack := stack[:]
		stack = []s{}
		for i := 0; i < len(newStack); i++ {
			if newStack[i].p1 == nil && newStack[i].p2 == nil {
				continue
			} else if newStack[i].p1 != nil || newStack[i].p2 != nil {
				return false
			} else if newStack[i].p1.Val != newStack[i].p2.Val {
				return false
			} else {
				stack = append(stack, s{
					p1: stack[i].p1.Left,
					p2: stack[i].p2.Left,
				})
				stack = append(stack, s{
					p1: stack[i].p1.Right,
					p2: stack[i].p2.Right,
				})
			}
		}
	}
	return true
}
